Check dependencies

huangruizhe opened the pull request:
**Background**
------------------------------------------------------------------------------------
Checking dependencies (and suggesting what to do) actually depends on several factors:
1. which interpreter to use --- resulting in pip or pip3
2. location of site-packages --- /usr/local/lib/pythonX.Y/site-packages or ~/.local/lib/pythonX.Y/site-packages
3. package manager or virtual environment --- conda install or pip install or ...

The cases are complicated, but we can still provide a common, simple and light-weighted dependency-checking solution, among many, for un-advanced users to enable them to get the project to run as smoothly/quickly as possible.

In the configuration or trial process, it is important to prompt up immediate and informative error (about dependencies and other), if there is, and suggest possible solutions. Otherwise, the users become impatient. In addition, as Yiwen suggested, there should be another documentation about dependencies and suggested "pip install ...." commands.

**Design**
------------------------------------------------------------------------------------
**Based on the above principles, we consider the following three cases for dependency-checking and package installation suggestions:**

1. users with root access: install everything on system-wide (easiest)
2. users without root access: pip install --user package_name
3. introducing the package and env manager, such as conda: install miniconda first and use the interpreter and environment enclosed
4. virtualenv --- not tested
5. pipenv --- too slow

Not sure what are common practices for popular python projects.

**Therefore, the implementation of dependency checking process in this PR is as follows:**
1. check_python3
2. check_packages --> if passed, then ok
3. otherwise, suggest what packages/dependencies are required. Note that, the choice of method of installation is users' own freedom. Nevertheless, we can provide some suggestions here:
3.1 suggestions for root-user
3.2 suggestions for non-root-user
3.3 suggestions for users who would like to additionally install a miniconda

**Here shows the printed messages for a fresh start**
------------------------------------------------------------------------------------
```
$ ./run.sh
scripts/dependencies/check_packages.sh not all the required python packages are installed.
Packages required: tensorboard_logger torch>=0.4.0 torchvision>=0.2.1
According to your situation, please do as follows (from your experiment directory, e.g. egs/dbs2018/v1):

(1) If you have root access and would like to install system-wide packages:
. ./path.sh
pip3 install tensorboard_logger
pip3 install http://download.pytorch.org/whl/cu90/torch-0.4.0-cp35-cp35m-linux_x86_64.whl
pip3 install torchvision

(2) (Recommended) If you do not have root access and would like to install packages under the user scheme:
. ./path.sh
pip3 install --user tensorboard_logger
pip3 install --user http://download.pytorch.org/whl/cu90/torch-0.4.0-cp35-cp35m-linux_x86_64.whl
pip3 install --user torchvision

(3) If you would like to use miniconda3 as package manager:
. ./path.sh
. ./scripts/dependencies/miniconda3/install_miniconda3.sh
. ./scripts/dependencies/miniconda3/path.sh
conda install pytorch torchvision cuda90 -c pytorch
pip install tensorboard_logger
```

------------------------------------------------------------------------------------

References:

Pipenv
[1] http://docs.python-guide.org/en/latest/dev/virtualenvs/ 
[2] https://robots.thoughtbot.com/how-to-manage-your-python-projects-with-pipenv
[3] https://packaging.python.org/tutorials/managing-dependencies/

Mechanism of installing Python modules
pip install --user
[4] https://docs.python.org/3/install/index.html#inst-alt-install-user 
[5] https://docs.python.org/3/installing/index.html#installing-index
[6] https://stackoverflow.com/questions/7143077/how-can-i-install-packages-in-my-home-folder-with-pip

virtualenv
[6] https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments

huangruizhe commented:
It seems this mechanism is a bit complex -- might need to cut down some parts of it

huangruizhe commented:
In short, what has been done?

1. made the local/check_dependencies for icdar2015 working more cleanly
2. fixed all dependency-checking-related issues (connected the whole process which has traps previously) and permission issues (chmod +x)
3. deleted some now-unused scripts
4. discovered that miniconda3 is NOT necessary

Not been done:

1. check the dsb2018 experiment and compatibility

danpovey commented on scripts/dependencies/check_packages.sh:
```diff
@@ -23,26 +23,115 @@ parser = argparse.ArgumentParser(description="Checks if all required packages"
+                dependency_list_rest.remove(pkg)
+
+        suggestion += ". ./path.sh \n"
+        suggestion += ". ./scripts/dependencies/miniconda3/install_miniconda3.sh \n"
```
this file wouldn't need to be sourced, just run. i.e. just
`scripts/dependencies/miniconda3/install_miniconda3.sh`

huangruizhe replied:
Just realized the difference. Thanks!

danpovey commented on scripts/dependencies/miniconda3/path.sh:
```diff
@@ -0,0 +1,2 @@
+export PATH="${HOME}/miniconda3/bin:$PATH"
```
Separating out this part of the path into a separate file is not really ideal.
We don't rely on the user changing their path and then running the script; we rely on everything that the user will need being in path.sh.  So if they wanted to use miniconda they might, for instance, uncomment some stuff in path.sh.
That is important when they run jobs using queueing software via wrappers like run.pl or queue.pl.

huangruizhe replied:
Got it. Shall we keep this part back in the original path.sh and comment it out?

danpovey commented on scripts/dependencies/check_packages.sh:
```diff
@@ -23,26 +23,115 @@ parser = argparse.ArgumentParser(description="Checks if all required packages"
 parser.add_argument('--requirements_path', type=str, 
                     default="scripts/dependencies/requirements.txt",
                     help='Path of the downloaded requirements file')
-args = parser.parse_args()
```
this shouldn't have the .sh suffix if it is a python script.  (this was a pre-existing issue).

danpovey commented:
@huangruizhe, I spoke with Yiwen and he does not object if we just remove the miniconda stuff.
I suggest that you work on a version of your dependency-checking script that removes the miniconda dependency and just uses pip3. 

But be sure to make the changes in all the recipes, not just the icdar one.  You don't have to test them all, others can do that.

huangruizhe commented:
Got it. Have made the following changes:
1. fixed the file names and some commands, as pointed out
2. removed the miniconda stuff
3. made the changes of dependencies checking in all the recipes
4. added support for local requirement.txt

Issues to be determined:
1. Should I keep env settings of miniconda3 in path.sh and comment them out? (Now, they are kept and commented out)
2. Should we keep the miniconda3 and pytorch installation scripts? They are not used for the moment.
3. The developer of each recipe might need to check the correctness of dependency-checking tool. E.g. I have updated the path.sh in the dsb2018 recipe.

Sample printed messages for a fresh start:

```
$ ./run.sh
scripts/dependencies/check_packages.py: Not all the required python packages are installed.
Packages required: tensorboard_logger torch>=0.4.0 torchvision>=0.2.1
Please do as follows (from your experiment directory, e.g. egs/dbs2018/v1):

pip3 install --user tensorboard_logger
pip3 install --user http://download.pytorch.org/whl/cu90/torch-0.4.0-cp35-cp35m-linux_x86_64.whl
pip3 install --user torchvision

If you have root access and would like to install system-wide packages, just remove the "--user" option.
```

huangruizhe commented:
Fixed.

danpovey commented:
Thanks.  If there are no objections soon (Yiwen?)  I'll merge this.

YiwenShaoStephen commented:
No problem. It looks good. Thanks! @huangruizhe

ChunChiehChang commented:
Should there be a check for matplotlib? The module is used in scripts/waldo/data_visualization.

danpovey commented:
Good idea.

On Thu, May 24, 2018 at 6:58 PM, ChunChiehChang <notifications@github.com>
wrote:

> Should there be a check for matplotlib? The module is used in
> scripts/waldo/data_visualization.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/waldo-seg/waldo/pull/39#issuecomment-391891240>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu0joNFawg00sFBTgEDcgySY6AC5Sks5t1zsjgaJpZM4UHJ4Y>
> .
>
